src/foam/core/FOAMException.js

```diff
@@ -33,48 +57,130 @@ foam.CLASS({
     setMessage_(message);
     getHostname();
   }
+
+  public FOAMException(String message, String errorCode, Throwable cause) {
+    super(message, cause);
+    setMessage_(message);
+    setErrorCode(errorCode);
+    getHostname();
+  }
         `);
       }
     }
   ],
 
   properties: [
     {
-      name: 'javaGenerateConvenienceConstructor',
-      value: false,
-      transient: true,
-      visibility: 'HIDDEN'
+      name: 'exceptionMessage',
+      class: 'String',
+      value: '{{message_}}',
+      visibility: 'RO'
     },
     {
       name: 'message_',
       class: 'String',
       visibility: 'RO'
     },
+    {
+      name: 'errorCode',
+      class: 'String',
+      visibility: 'RO'
+    },
     {
       name: 'hostname',
       class: 'String',
       javaFactory: 'return System.getProperty("hostname", "localhost");',
       visibilty: 'RO'
     }
   ],
-  
+
+  // templates: [
```

Comment:
@yij793 was this in your original design?

Reply:
@jlhughes Yeah, but before I use Messages instead of property. Your method is much flexible